File: SecondYearCS/OOP/CivitasRB/lib/SorpresaSalirCarcel.rb
require_relative 'sorpresa.rb'
module CivitasRB
  class SorpresaSalirCarcel < Sorpresa
    #Constructores por defecto y con parametro asdfas
    def initialize(mazo,texto)
      super(texto)
      @mazo = mazo
    end

    def aplicarAJugador_salirCarcel(act,todos)
      if(self.jugadorCorrecto(act, todos))
        self.informe(act,todos)
        tieneSC = faflse
        todos.each do |jugador|
          if jugador.tieneSalvoconducto
            tieneSC = true
          end
        end
        if !tieneSC
          todos.at(act).obtenerSalvoConducto(self)
          self.salirDelMazo()
        end
      end
    end
    
    def salirDelMazo
        @mazo.inhabilitarCartaEspecial(self)
    end
    
    def usada
        @mazo.habilitarCartaEspecial(self)
    end
    
    def to_s
      return super.to_s
    end
    
  end
end

File: SecondYearCS/OOP/Civitas/src/civitas/SorpresaSalirCarcel.java
package civitas;

import java.util.ArrayList;

public class SorpresaSalirCarcel extends Sorpresa{
    
    private MazoSorpresa Mazo;
    
    //se utiliza para evitar la carcel
    SorpresaSalirCarcel(MazoSorpresa mazo, String texto){
        super(texto);
        this.Mazo = mazo;
        Mazo.habilitarCartaEspecial(this);
    }
    
    @Override
    public void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        if( this.jugadorCorrecto(actual, todos) ){
            informe(actual,todos);
            boolean tieneSC = false;
            for( int i = 0; i < todos.size() && !tieneSC; i++ ){
                if( todos.get(i).tieneSalvoconducto() ){
                    tieneSC = true;
                }
            }
            if( !tieneSC ){
                todos.get(actual).obtenerSalvoConducto(this);
                salirDelMazo();
            }
        }
    }

    void salirDelMazo(){
        this.Mazo.inhabilitarCartaEspecial(this);
    }
    
    void usada(){
        this.Mazo.habilitarCartaEspecial(this);
    }

    @Override
    public String toString(){
       return super.toString();
    }

}